### ISSUE TITLE
Concurrent requests to sync routes randomly hang since 0.15.0

### ISSUE THREAD

**Issue by jlaine** [Kludex/starlette#1248]
### Checklist

- [x] The bug is reproducible against the latest release and/or `master`.
- [x] There are no similar issues or pull requests to fix it yet.

### Describe the bug

`starlette` is used for the demo application + server of the `aioquic` QUIC + HTTP/3 stack. Since starlette 0.15.0 (i.e. also with 0.16.0) I have noticed that concurrent requests (on a single connection) randomly hang, before even entering the route. 

NOTE: I am not claiming this is definitely a bug in `starlette`, but I am having a hard time debugging this issue, and rolling back to startlette 0.14.2 fixes it so it may be related to the `anyio` changes.

### To reproduce

I have observed this on routes as simple as:

```
def padding(request):
    return PlainTextResponse("Z" * 5000)
```

The key here is that this is *not* a coroutine. If I add the `async` keyword the route seems to work reliably.

I have also observed the same behaviour with the built-in "Not Found" handler: `ExceptionMiddleware.http_exception`. As for the previous case, changing this method to a coroutine fixes the problem.

The common feature I have found so far with the routes that hang is that the code gets invoked via `run_in_threadpool`.

### Environment

- OS: Linux
- Python version: 3.9.2
- Starlette version: 0.16.0

**Comment by jlaine** [comment]
NOTE: when I say *concurrent*, I do really mean concurrent on the same socket. HTTP/3 does not serialize request/responses (like HTTP/1.1) nor does it suffer from head-of-line blocking.

**Comment by jlaine** [comment]
This might actually be an issue inside `anyio`'s asyncio backend.

The following is a minimal reproduction which triggers the condition. It doesn't even need requests, simply running `run_in_threadpool` can lead to a hang. The following code works about half the time for me, and hangs the other times. If you don't get a hang try increasing the number of iterations.

```python
import asyncio
import random
import time

from starlette.concurrency import run_in_threadpool


def myfunc(*args, **kwargs):
    print("myfunc")
    time.sleep(random.random() * 0.1)


async def myapp():
    await asyncio.sleep(random.random())
    await run_in_threadpool(myfunc)
    await asyncio.sleep(random.random())


loop = asyncio.get_event_loop()
tasks = [asyncio.ensure_future(myapp()) for x in range(20)]
loop.run_until_complete(asyncio.gather(*tasks))
```

This seems to be something to do with idle worker reuse, possibly due to a mismatch between the initial "root task" and the one for which the worker gets repurposed. If I change the following line in anyio to an `if True` things work (at the expense of infinite workers):

https://github.com/agronholm/anyio/blob/a024b29ac23a74b9ee8d1042c49865a86fb6040e/src/anyio/_backends/_asyncio.py#L781

**Comment by jlaine** [comment]
After further investigation this really looks like a bug in `anyio`. I'd prefer if this issue was kept around until a fix lands in `anyio` so that `starlette` can bump its dependency version, but if you feel this is not the right place you are of course free to close this issue!

**Comment by JayH5** [comment]
Thanks for pointing this out. Happy for the ticket to stay open for discoverability. Do you have an issue in anyio you can link to?

**Comment by jlaine** [comment]
Sure it's issue https://github.com/agronholm/anyio/issues/323 and @agronholm has kindly released `anyio` 3.3.0 with a fix for the issue.
